utils/warnings.ml

```diff
@@ -631,6 +633,9 @@ let message = function
          which shadows the existing one.\n\
          Hint: Did you mean 'type %s = unit'?" name
   | Unused_functor_parameter s -> "unused functor parameter " ^ s ^ "."
+  | Match_on_mutable_will_allocate_closure ->
+    "This pattern-matching captures a mutable arguments in a curried function.\n\
```

Comment:
Possibly people will quibble with the description of a lazy match as "capturing a mutable argument". We might want to adjust the wording so that it makes sense for both matches on mutable fields and lazy matches. Perhaps something like: " This pattern depends on mutable state. It prevents the remaining arguments from being uncurried, which will cause additional closure allocations."

Reply:
Updated